Improve HIP documentation (#314)
nlesc-recruit/cudawrappers: README.md

## Usage

> CMake is used to build cudawrappers. To compile the library, run:

Reviewer: As there is nothing to compile:
```suggestion
CMake is used to build cudawrappers. To build the library, run:
```

make -C build
```

This creates a `build` directory. Since cudawrappers is header-only, no library objects will be built. 
For more details on the building requirements and testing, see the [developer documentation](README.dev.md).

To install cudawrappers at a specific location (e.g., ~/.local), use:
```shell
cmake -DCMAKE_INSTALL_PREFIX=$HOME/.local -S . -B build
make -C build
make -C build install
```

### Enabling HIP Support

To enable HIP:

1. Add the following code to your project's `CMakeLists.txt` file:

    ```cmake
    enable_language(HIP)
    set(CUDAWRAPPERS_BACKEND "HIP")
    ```
    Alternatively, use the `-DCUDAWRAPPERS_BACKEND=HIP` argument when running CMake.

2. Ensure that every target your project builds is 'hipfied'. This is done by marking relevant source files as 'HIP' compatible:

    ```cmake
    set_source_files_properties(source.cpp PROPERTIES LANGUAGE HIP)
    ```

3. Optionally, use `#ifdef (__HIP__)` directives in your source code to enable/disable certain sections for HIP.

4. **Build** while ensuring that the `hipcc` compiler is selected. This can be done via the command line:

    ```shell
    CXX=hipcc cmake -B build
    ```

> **Note**: When building for both NVIDIA and AMD HIP, using **seperate** build folders (e.g, `build_nvidia` and `build_amd`) is encouraged. Additionally, please note that HIP deprecated support for so-called GPU 'contexts'. Therefore, the usage of cudawrappers' `cu::Context` is discouraged and marked deprecated since cudawrappers version `0.9.0`.

Reviewer 2: Here is says it is discouraged but context just dont work on cudawrappers when you use hip. You get a silent failure with an empty class...

Author: Done

make -C build
```

This creates a `build` directory. Since cudawrappers is header-only, no library objects will be built.
For more details on the building requirements and testing, see the [developer documentation](README.dev.md).

> To install cudawrappers at a specific location (e.g., ~/.local), use:

Reviewer: ```suggestion
To install cudawrappers to a specific location (e.g., ~/.local), use:
```

```shell
cmake -DCMAKE_INSTALL_PREFIX=$HOME/.local -S . -B build
make -C build
make -C build install
```

### Enabling HIP Support

To enable HIP:

1. Add the following code to your project's `CMakeLists.txt` file:

    ```cmake
    enable_language(HIP)
    set(CUDAWRAPPERS_BACKEND "HIP")
    ```

    Alternatively, use the `-DCUDAWRAPPERS_BACKEND=HIP` argument when running CMake.

2. Ensure that every target your project builds is 'hipfied'.
    This is done by marking relevant source files as 'HIP' compatible:

    ```cmake
    set_source_files_properties(source.cpp PROPERTIES LANGUAGE HIP)
    ```

3. Optionally, use `#ifdef (__HIP__)` directives in your source code to enable/disable certain sections for HIP.

> 4. **Build** while ensuring that the `hipcc` compiler is selected.

Reviewer: ```suggestion
4. **Build** ensure that the `hipcc` compiler is selected.
```

Note: is this still necessary at all?

    This can be done via the command line:

    ```shell
    CXX=hipcc cmake -B build
    ```

**Note**: When building for both NVIDIA and AMD HIP,
using **seperate** build folders (e.g, `build_nvidia` and `build_amd`) is encouraged.

> Additionally, please note that contrary to CUDA, HIP does not support GPU 'contexts'.

Reviewer: HIP still supports it, but has it marked deprecated. Not sure it's worth changing this to reflect  that though.

Author: Originally, it was written in the way you are suggesting, however, Mattia pointed out that HIP basically constructs empty code blocks. Therefore, the wording was changed.

Reviewer: cudawrappers does that indeed, my point was more aimed at HIP itself, which does still support contexts although they are deprecated.

Reviewer: I would change "HIP" to "The HIP backend" to indicate that it relates to cudawrappers, not HIP itself. Looks good otherwise!
